### PR TITLE
8269661: JNI_GetStringCritical does not lock char array

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this fix to jni_GetStringCritical.  When object pinning is
being used it pins/unpins the String argument, but that's the wrong object.
It should be pinning the associated byte array (when !latin1), as it is a
reference into that array that will be returned and must be imobilized.

Presently only Shenandoah is affected by this bug, as it is the only GC that
uses object pinning (at the region level) rather than the GC locker.  But
both G1 and ZGC plan to use object pinning in the future.  But with region
pinning the problem is still going to be rare because the String and its
value array are likely to be allocated in the same region.

In addition, if pinning the String's value array, we also need to prevent
string deduplication from changing the array while in the critical section.
Otherwise, the unpin when the critical section is released will use the
wrong object for the unpin operation.  We accomplish this by marking the
String as no longer subject to string deduplication.  As that state is
sticky, when pinning is used a String can't be deduplicated after being used
in a critical section.  We could add a critical section counter to String,
but for now we're going to guess that isn't worth the effort.  (String has
at least 2 bytes available for the two string deduplication flags plus such
a counter (more in some configurations, but always at least 2 at present)).

As part of the restructuring to accomplish these changes, it proved to be
easy to avoid using the GC locker (or pinning) for the latin1 case, where a
copy of the value array will be used.  Hence this change also addresses
JDK-8269650.

Testing:
Existing tests for GetStringCritical are in various vmTestbase/gc/lock
tests and vmTestbase/nsk/stress/jni/jnistress004.java.

Ran mach5 tier1, tier5.  Tier5 is where the vmTestbase/gc/lock and
vmTestbase/nsk/stress/jni tests are run.

Locally (linux-x64) ran those tests with Shenandoah, with and without
string deduplication enabled.

/issue add JDK-8269650

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8269661](https://bugs.openjdk.java.net/browse/JDK-8269661): JNI_GetStringCritical does not lock char array
 * [JDK-8269650](https://bugs.openjdk.java.net/browse/JDK-8269650): Optimize gc-locker in [Get|Release]StringCritical for latin string

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to ea8bd3861bfbf44626b3bfc72c60413baeda2873
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to ea8bd3861bfbf44626b3bfc72c60413baeda2873

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/209/head:pull/209` \
`$ git checkout pull/209`

Update a local copy of the PR: \
`$ git checkout pull/209` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/209/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 209`

View PR using the GUI difftool: \
`$ git pr show -t 209`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/209.diff">https://git.openjdk.java.net/jdk17/pull/209.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Hi Kim,

Based on  your description and previous discussion this all looks good to me.

Thanks,
David

**Review by tschatzl — APPROVED** [review]

File: src/hotspot/share/prims/jni.cpp [hunk 1]
```diff
@@ -2864,6 +2891,7 @@ JNI_ENTRY(const jchar*, jni_GetStringCritical(JNIEnv *env, jstring string, jbool
       }
       ret[s_len] = 0;
     }
+    if (isCopy != NULL) *isCopy = JNI_TRUE;
```
[the remark this answers is not in the document]
Comment (kimbarrett) [reply src/hotspot/share/prims/jni.cpp:2894]: I noticed that, but decided against making any change to that behavior. I think it's arguable whether it's correct usage, but I can imagine a caller that used a JNI_FALSE isCopy result to indicate no copy was attempted so no OOME could have happened so the NULL check of the result isn't needed.  The specification doesn't say what the isCopy value should be in the OOME case.

**Comment by kimbarrett** [comment]
Thanks for reviews @dholmes-ora and @tschatzl .

**Pull request closed without merging**